woocommerce/google-listings-and-ads
# Integrate phone number UI with verification APIs

### Changes proposed in this Pull Request:

Part of #962.

- Remove the update of phone number from the `updateGoogleMCContactInformation()` action.
- Add actions to communicate with phone verification APIs.
- Integrate `<VerifyPhoneNumberContent>` with phone verification APIs.

### ❗ Important

- In this PR, the phone number UI communicates with Google's API on production env. **It will really send the verification code to the phone number via SMS/phone call. Please make sure the entered number is your own or someone you know.**
- There's a rate limit on Google's verification API. A maximum of five attempts to verify the same phone number every four hours.

### Screenshots:

💡  To avoid exposing my personal phone number on public network 😅 , I temporarily tampered with the UI display on my local during the recording, so the phone number inside is not real. 🙈 

#### 🎥 . Verify phone number via SMS

![Kapture 2021-09-24 at 15 31 48](https://user-images.githubusercontent.com/17420811/134638761-a29b03cc-486e-4ba8-af83-f08f39ec1b86.gif)

#### 🎥 . Errors handling

![Kapture 2021-09-24 at 15 39 27](https://user-images.githubusercontent.com/17420811/134638804-b99038dd-a0d7-4d8a-88b5-c0548413e5ea.gif)

#### 🎥 . A verification code sent via phone call

https://user-images.githubusercontent.com/17420811/134638842-d130b546-7b0a-4cfe-8ba2-8115ade47a78.mp4

#### 🎥 . A verification code sent via SMS

https://user-images.githubusercontent.com/17420811/134638880-e68274b7-68fd-41a6-ba13-498e6fc5708e.mp4

### Detailed test instructions:

💡  Consider that there's a rate limit. It recommends testing in the below order to avoid not being able to test for a long time after reaching the rate limit.

#### Test the errors handling when sending back the verification code

1. Go to the edit phone number section in the Settings page
1. Start a phone number verification via SMS
1. Intentionally verify with another incorrect code to see if the error message shows up
1. Don't leave this page and stay in the state of entering the verification code, wait for 15 minutes and then come back to verify again. It should show up an error message that the code has expired.

#### Test the API integration

1. Go to the [Google Merchant Center](https://merchants.google.com/mc/merchantprofile/businessinfo) and save an unverified number
    ![2021-09-24 16 32 56](https://user-images.githubusercontent.com/17420811/134644111-80d7151c-eb78-4121-b23c-84189f7e1c64.png)
1. Start a phone number verification via SMS
1. Enter the received verification code to see if it passes the verification and then also updating to Google Merchant Center as a verified number
1. Repeat this test from step 1 again by getting a code via phone call

#### Test the errors handling when requesting the verification code

1. Request verification code over five times with the same phone number
1. It should show up an error message about the rate limit

### Changelog entry

> Add - Verify user's phone number via SMS or phone call at the last step of the onboarding flow and on the settings page. And update the verified phone number to user's connected Google Merchant Center account.
